src/Microsoft.AspNetCore.Routing/Matching/EndpointComparer.cs

```diff
@@ -0,0 +1,162 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using Microsoft.AspNetCore.Http;
+
+namespace Microsoft.AspNetCore.Routing.Matching
+{
+    // Use to sort and group Endpoints. RouteEndpoints are sorted before other implementations.
+    //
+    // NOTE:
+    // When ordering endpoints, we compare the route templates as an absolute last resort.
+    // This is used as a factor to ensure that we always have a predictable ordering
+    // for tests, errors, etc.
+    //
+    // When we group endpoints we don't consider the route template, because we're trying
+    // to group endpoints not separate them.
+    //
+    // TLDR: 
+    //  IComparer implementation considers the template string as a tiebreaker.
+    //  IEqualityComparer implementation does not.
+    //  This is cool and good.
+    internal class EndpointComparer : IComparer<Endpoint>, IEqualityComparer<Endpoint>
```

Comment:
Yeah this was the one part I was a little unsure what to do about. The built in comparers look a route-endpoint specific stuff.